das/aggregator.go

```diff
@@ -0,0 +1,218 @@
+//
+// Copyright 2022, Offchain Labs, Inc. All rights reserved.
+//
+
+package das
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/bits"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/offchainlabs/nitro/arbstate"
+	"github.com/offchainlabs/nitro/blsSignatures"
+)
+
+type AggregatorConfig struct {
+	assumedHonest   int
+	retentionPeriod time.Duration
+}
+
+type Aggregator struct {
+	config   AggregatorConfig
+	services []serviceDetails
+
+	/// calculated fields
+	requiredServicesForStore       int
+	maxAllowedServiceStoreFailures int
+}
+
+type serviceDetails struct {
+	service    DataAvailabilityService
+	pubKey     blsSignatures.PublicKey
+	signerMask uint64
+}
+
+func NewAggregator(config AggregatorConfig, services []serviceDetails) *Aggregator {
+	return &Aggregator{
```

Comment:
Or do we want to allow one DAS to incorporate multiple signers (presumably because it is doing some signature aggregation internally, and we want to count those as separate signers for retrievability purposes?

Reply:
We talked about nesting aggregators and I think we decided against it at least for this version due to the complexity it would introduce, so I added in the check for the signer mask just having 1 bit.